elk/extraction/extraction.py

```diff
@@ -95,30 +95,166 @@ def explode(self) -> list["Extract"]:
+    But for now, we need to use a threadpool utilize all the InferenceServer workerss
+    """
+    ranks_and_splits: list[ExtractHiddenThreadParam] = [
+        ExtractHiddenThreadParam(
+            rank=device_rank,
+            world_size=len(server.devices),
+            split_type=split_name,
+            device=device,
+        )
+        for device_rank, device in enumerate(server.devices)
+        for split_name in split_names
+    ]
+    print("Rank and splits: ", ranks_and_splits)
+    # 2 threads per device - This is so that the workers of the
+    # InferenceServer should be fully saturated.
+    tp_size = len(ranks_and_splits)
+    with ThreadPoolExecutor(max_workers=tp_size) as executor:
+        hiddens = map_threadpool(
+            items=ranks_and_splits,
+            # TODO: Probably can speed this up by loading the prompts first,
+            # then passing them in here.
+            func=lambda param: extract_hiddens(
+                cfg=cfg,
+                server=server,
+                device=param.device,
+                split_type=param.split_type,
+                world_size=param.world_size,
+                rank=param.rank,
+            ),
+            threadpool=executor,
+        )
+    flattened_hiddens = flatten_list(hiddens)
+    return extracted_hiddens_to_dataset_with_name(
+        extracted_hiddens=flattened_hiddens,
+        ds_name=ds_name,
+    )
+
+
+@dataclass
+class SmallerOutput:
+    hidden_states: torch.Tensor | None
+    lm_logits: torch.Tensor
+
+
+def func_to_run(
+    model_output: ModelOutput,
+    worker_device: torch.device,
+    has_lm_preds: bool,
+    tokens_shared: torch.Tensor,
+    answer_len: int,
+    layer_indices: tuple[int, ...],
+    token_loc: Literal["first", "last", "mean"],
+) -> SmallerOutput:
+    if has_lm_preds:
+        output_logits = model_output.logits[..., -answer_len:, :]
+        log_p = output_logits.log_softmax(dim=-1)
+        tokens_on_device = tokens_shared.to(worker_device)
+        logit_gathered = log_p.gather(-1, tokens_on_device).sum().to("cpu")
+        returned_logits = pytree_map(lambda x: x.cpu().share_memory_(), logit_gathered)
+    else:
+        returned_logits = None
+
+    returned_hiddens = (
+        model_output.get("decoder_hidden_states") or model_output["hidden_states"]
+    )
+    # Throw out layers we don't care about
+    hiddens = [returned_hiddens[i] for i in layer_indices]
+
+    # Current shape of each element: (batch_size, seq_len, hidden_size)
+    if token_loc == "first":
+        hiddens = [h[..., 0, :] for h in hiddens]
+    elif token_loc == "last":
+        hiddens = [h[..., -1, :] for h in hiddens]
+    elif token_loc == "mean":
+        hiddens = [h.mean(dim=-2) for h in hiddens]
+    else:
+        raise ValueError(f"Invalid token_loc: {token_loc}")
+    converted_hiddens = pytree_map(lambda x: x.cpu().share_memory_(), hiddens)
+
+    return SmallerOutput(lm_logits=returned_logits, hidden_states=converted_hiddens)
```

Comment:
the function to run on the server.  note that you need to shift the shared tensor to the worker's device porperly